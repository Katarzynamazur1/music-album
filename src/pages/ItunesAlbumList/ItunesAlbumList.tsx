import { useEffect, useState } from "react";
import Album from "../../common/musicAlbum/Album";
import CircleLoader from "../../common/CircleLoader";
import PageTitle from "../../common/PageTitle";
import {
  getAlbumsFromItunesAlbumData,
  ItunesAlbumDataEntry,
} from "./itunesDataTransformer";
import { fetchSalesIfNotPresent } from "../../state/sales/salesActions";
import { useAppDispatch, useAppSelector } from "../../state/stateHooks";
import { getDiscountForItem } from "../../dataMocking/salesDataMock";

const addDelay = () => {
  return new Promise((resolve) =>
    setTimeout(() => {
      resolve(null);
    }, 1000)
  );
};

export interface ItunesTopAlbumsResponseData {
  feed: {
    entry: ItunesAlbumDataEntry[];
  };
}

const ItunesAlbumList = () => {
  const [albumDataEntries, setAlbumDataEntries] = useState<
    ItunesAlbumDataEntry[] | null
  >(null);

  const dispatch = useAppDispatch();
  const salesInformation = useAppSelector((state) => state.sales.data);

  useEffect(() => {
    const fetchAndSetAlbums = async () => {
      const fetchResponse = await fetch(
        "https://itunes.apple.com/us/rss/topalbums/limit=100/json"
      );
      const fetchedData =
        (await fetchResponse.json()) as ItunesTopAlbumsResponseData;

      setAlbumDataEntries(fetchedData.feed.entry);
    };

    fetchAndSetAlbums();
    dispatch(fetchSalesIfNotPresent);
  }, []);

  const albumData = albumDataEntries
    ? getAlbumsFromItunesAlbumData(albumDataEntries)
    : [];

  const albumComponents = albumData?.map((album) => {
    const salesItem = {
      id: album.id,
      value: album.description.price,
    };
    const discountForAlbum = getDiscountForItem(salesItem, salesInformation);

    return (
      <Album
        id={album.id}
        key={album.id}
        number={album.number}
        coverImageUrl={album.coverImageUrl}
        description={album.description}
        discountValue={discountForAlbum}
      />
    );
  });

  return (
    <>
      <PageTitle title="Top Albums" />
      <CircleLoader show={albumDataEntries === null} />
      {albumComponents}
    </>
  );
};

export default ItunesAlbumList;
