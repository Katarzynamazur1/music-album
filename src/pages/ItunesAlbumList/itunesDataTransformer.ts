import { AlbumProps } from "../../common/musicAlbum/Album";

interface ObjectWithlabel {
  label: string;
}

interface PriceObject {
  label: string;
  attributes: {
    amount: string;
  };
}

interface ItunesAlbumIdObject {
  attributes: {
    "im:id": string;
  };
}

export interface ItunesAlbumDataEntry {
  id: ItunesAlbumIdObject;
  "im:name": ObjectWithlabel;
  "im:artist": ObjectWithlabel;
  "im:price"?: PriceObject;
  "im:image"?: ObjectWithlabel[];
  "im:releaseDate"?: ObjectWithlabel;
}

export const getAlbumsFromItunesAlbumData = (
  itunesAlbumEntries: ItunesAlbumDataEntry[]
): AlbumProps[] => {
  const transformedData = itunesAlbumEntries.map((itunesAlbum, index) => {
    const priceValueFromItunes = itunesAlbum?.["im:price"]?.attributes.amount;

    return {
      id: itunesAlbum.id.attributes["im:id"],
      number: index + 1,
      coverImageUrl: itunesAlbum?.["im:image"]?.[2]?.label || "",
      description: {
        title: itunesAlbum["im:name"].label,
        artist: itunesAlbum["im:artist"].label,
        price: Number(priceValueFromItunes),
        releaseDate: itunesAlbum["im:releaseDate"]?.label,
      },
    };
  });

  return transformedData;
};
