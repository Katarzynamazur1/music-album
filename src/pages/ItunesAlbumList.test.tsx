import { configureStore } from "@reduxjs/toolkit";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import rootReducer from "../state/reducers";
import ItunesAlbumList from "./ItunesAlbumList/ItunesAlbumList";

const storeWithoutLogger = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) => getDefaultMiddleware(),
});

describe("ItunesAlbumList.tsx", () => {
  it("Matches snapshot when api data is still loading", () => {
    const { container } = render(
      <Provider store={storeWithoutLogger}>
        <ItunesAlbumList />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it("loaded appropriate albums when api data has been loaded", () => {
    const { container } = render(
      <Provider store={storeWithoutLogger}>
        <ItunesAlbumList />
      </Provider>
    );

    // wait for response to arrive
    // Exercise:  How can we wait for render result to contain albums instead of the loader? How to do this?
    //          Q: Which library can provide this functionality RTL or Jest?

    // loader is no longer present
    // appropriate albums were loaded
    // expect ?
  });
});
