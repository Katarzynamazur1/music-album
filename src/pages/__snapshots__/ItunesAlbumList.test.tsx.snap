// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItunesAlbumList.tsx Matches snapshot when ? 1`] = `
<div>
  <div
    style="margin-top: 30px; margin-bottom: 40px; text-align: center; font-family: 'Courier New', monospace; color: rgba(255, 255, 255, 0.7);"
  >
    Top Albums
  </div>
  <div
    style="text-align: center; margin-top: 100px;"
  >
    <div
      class="loader"
    />
  </div>
</div>
`;

exports[`ItunesAlbumList.tsx Matches snapshot when api data is not loaded yet 1`] = `
<div>
  <div
    style="margin-top: 30px; margin-bottom: 40px; text-align: center; font-family: 'Courier New', monospace; color: rgba(255, 255, 255, 0.7);"
  >
    Top Albums
  </div>
  <div
    style="text-align: center; margin-top: 100px;"
  >
    <div
      class="loader"
    />
  </div>
</div>
`;

exports[`ItunesAlbumList.tsx Matches snapshot when api data is still loading 1`] = `
<div>
  <div
    style="margin-top: 30px; margin-bottom: 40px; text-align: center; font-family: 'Courier New', monospace; color: rgba(255, 255, 255, 0.7);"
  >
    Top Albums
  </div>
  <div
    style="text-align: center; margin-top: 100px;"
  >
    <div
      class="loader"
    />
  </div>
</div>
`;
